instrumentation/rabbitmq-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/rabbitmq/RabbitChannelInstrumentation.java

```diff
@@ -92,64 +93,85 @@ public void transform(TypeTransformer transformer) {
-      CURRENT_RABBIT_CONTEXT.set(context);
-      helper().setChannelAndMethod(context, request);
-      scope = context.makeCurrent();
-    }
+        if (!channelInstrumenter(request).shouldStart(parentContext, request)) {
+          return new ChannelMethodAdviceScope(callDepth, null, null, null);
+        }
 
-    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
-    public static void stopSpan(
-        @Advice.Thrown Throwable throwable,
-        @Advice.Local("otelCallDepth") CallDepth callDepth,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope,
-        @Advice.Local("otelRequest") ChannelAndMethod request) {
-      if (callDepth.decrementAndGet() > 0) {
-        return;
+        Context context = channelInstrumenter(request).start(parentContext, request);
+        CURRENT_RABBIT_CONTEXT.set(context);
+        helper().setChannelAndMethod(context, request);
+
+        return new ChannelMethodAdviceScope(callDepth, context, context.makeCurrent(), request);
       }
-      if (scope == null) {
-        return;
+
+      public void end(Throwable throwable) {
+        if (callDepth.decrementAndGet() > 0) {
+          return;
+        }
+        if (scope == null) {
+          return;
+        }
+
+        scope.close();
+
+        CURRENT_RABBIT_CONTEXT.remove();
+        channelInstrumenter(request).end(context, request, null, throwable);
       }
+    }
 
-      scope.close();
+    @Advice.OnMethodEnter
+    public static ChannelMethodAdviceScope onEnter(
+        @Advice.This Channel channel, @Advice.Origin("Channel.#m") String method) {
+      return ChannelMethodAdviceScope.start(CallDepth.forClass(Channel.class), channel, method);
+    }
 
-      CURRENT_RABBIT_CONTEXT.remove();
-      channelInstrumenter(request).end(context, request, null, throwable);
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void stopSpan(
+        @Advice.Thrown @Nullable Throwable throwable,
+        @Advice.Enter ChannelMethodAdviceScope adviceScope) {
+      adviceScope.end(throwable);
     }
   }
 
   @SuppressWarnings("unused")
   public static class ChannelPublishAdvice {
 
+    @Advice.AssignReturned.ToArguments(@Advice.AssignReturned.ToArguments.ToArgument(4))
     @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void setSpanNameAddHeaders(
+    public static AMQP.BasicProperties setSpanNameAddHeaders(
         @Advice.Argument(0) String exchange,
         @Advice.Argument(1) String routingKey,
-        @Advice.Argument(value = 4, readOnly = false) AMQP.BasicProperties props,
+        @Advice.Argument(4) AMQP.BasicProperties props,
         @Advice.Argument(5) byte[] body) {
       Context context = Java8BytecodeBridge.currentContext();
       Span span = Java8BytecodeBridge.spanFromContext(context);
+      AMQP.BasicProperties modifiedProps = props;
```

Comment:
It would have taken less renaming if you had renamed the method argument to `originalProps` and here used `AMQP.BasicProperties props = originalProps;`. I know that with inline advice you can't reassign read only arguments. Just curious did it fail with the original code when `props` was reassigned?

Reply:
I tried it out and got a slightly different exception. My understanding is that writing to method parameters is fine with non-inline advice. The advice code runs in a separate class and even when you change the value for the local parameter it isn't reflected in the original method. Where things break is when the same advice is applied as inline advice. There byte-buddy will detect the attempt to change the value of the method parameter and reject it because the parameter is read only.